File: 94-binary-tree-inorder-traversal/binary-tree-inorder-traversal.java
//MORRIS TRAV-O(n)
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> arr = new ArrayList<>();
        TreeNode curr=root;
        while(curr!=null){
            if(curr.left!=null){//find predeccesor
                TreeNode pred=curr.left;
                while(pred.right!=null && pred.right!=curr){
                    pred=pred.right;
                }
                if(pred.right==null){//conect pred's right' w curr
                    pred.right=curr;
                    curr=curr.left;
                }if(pred.right==curr){//remove link bw pred n curr
                    pred.right=null;
                    arr.add(curr.val);
                    curr=curr.right;
                }
            }
            else{//curr.left==null, no pred
                arr.add(curr.val);
                curr=curr.right;
            }
        }
        return arr;
    }
}

// // //rec
// // class Solution {
// //     List<Integer> ans = new ArrayList<>();
// //     public List<Integer> inorderTraversal(TreeNode root) {
// //         if (root == null) return ans;
// //         inorderTraversal(root.left);
// //         ans.add(root.val);
// //         inorderTraversal(root.right);
// //         return ans;
// //     }
// // }

// //iter-using stack
// class Solution {
//     public List<Integer> inorderTraversal(TreeNode root) {
//         List<Integer> list = new ArrayList<Integer>();

//         Stack<TreeNode> stack = new Stack<TreeNode>();
//         TreeNode cur = root;

//         while(cur!=null || !stack.empty()){
//             while(cur!=null){
//                 stack.add(cur);
//                 cur = cur.left;
//             }
//             cur = stack.pop();
//             list.add(cur.val);
//             cur = cur.right;
//         }

//         return list;
//     }
// }

File: 94-binary-tree-inorder-traversal/binary-tree-inorder-traversal.cpp
// class Solution {
// public:
//     vector<int> inorderTraversal(TreeNode* root) {
//         stack<TreeNode*> st;
//         TreeNode* node=root;
//         vector<int> inorder;
//         while(true){
//             if(node!=NULL){
//                 st.push(node);
//                 node=node->left;
//             }else{
//                 if(st.empty()==true) break;
//                 node=st.top();
//                 st.pop();
//                 inorder.push_back(node->val);
//                 node=node->right;
//             }
//         }
//         return inorder;
//     }
// };

//rec
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> nodes;
        inorder(root, nodes);
        return nodes;
    }
private:
    void inorder(TreeNode* root, vector<int>& nodes) {
        if (!root) {
            return;
        }
        inorder(root -> left, nodes);
        nodes.push_back(root -> val);
        inorder(root -> right, nodes);
    }
};


// //morris
// class Solution {
// public:
//     vector<int> inorderTraversal(TreeNode* root) {
//         vector<int> nodes;
//         while (root) {
//             if (root -> left) {
//                 TreeNode* pre = root -> left;
//                 while (pre -> right && pre -> right != root) {
//                     pre = pre -> right;
//                 }
//                 if (!pre -> right) {
//                     pre -> right = root;
//                     root = root -> left;
//                 } else {
//                     pre -> right = NULL;
//                     nodes.push_back(root -> val);
//                     root = root -> right;
//                 }
//             } else {
//                 nodes.push_back(root -> val);
//                 root = root -> right;
//             }
//         }
//         return nodes;
//     }
// };